src/neural/blas/winograd_convolution3.cc

```diff
@@ -0,0 +1,337 @@
+
+          for (int i = 0; i < kWinogradAlpha; i++) {
+            for (int j = 0; j < kWinogradAlpha; j++) {
+              if ((yin + i) >= 0 && (xin + j) >= 0 && (yin + i) < kHeight &&
+                  (xin + j) < kWidth) {
+                x[i][j] = input_channel[(yin + i) * kWidth + (xin + j)];
+              } else {
+                x[i][j] = 0.0f;
+              }
+            }
+          }
+
+          // Calculates transpose(B).x.B
+          // B = [[ 1.0,  0.0,  0.0,  0.0],
+          //      [ 0.0,  1.0, -1.0,  1.0],
+          //      [-1.0,  1.0,  1.0,  0.0],
+          //      [ 0.0,  0.0,  0.0, -1.0]]
+
+          //     WinogradTile T1, T2;
+
+          T1[0][0] = x[0][0] - x[2][0];
+          T1[0][1] = x[0][1] - x[2][1];
+          T1[0][2] = x[0][2] - x[2][2];
+          T1[0][3] = x[0][3] - x[2][3];
+          T1[1][0] = x[1][0] + x[2][0];
+          T1[1][1] = x[1][1] + x[2][1];
+          T1[1][2] = x[1][2] + x[2][2];
+          T1[1][3] = x[1][3] + x[2][3];
+          T1[2][0] = x[2][0] - x[1][0];
+          T1[2][1] = x[2][1] - x[1][1];
+          T1[2][2] = x[2][2] - x[1][2];
+          T1[2][3] = x[2][3] - x[1][3];
+          T1[3][0] = x[1][0] - x[3][0];
+          T1[3][1] = x[1][1] - x[3][1];
+          T1[3][2] = x[1][2] - x[3][2];
+          T1[3][3] = x[1][3] - x[3][3];
+
+          const auto V_incr = channels * kTiles * batch_size;
+          float* wTile_V = V_channel + channels * (block_y * kWtiles + block_x);
+
+          *wTile_V = T1[0][0] - T1[0][2];
+          wTile_V += V_incr;
+          *wTile_V = T1[0][1] + T1[0][2];
+          wTile_V += V_incr;
+          *wTile_V = T1[0][2] - T1[0][1];
+          wTile_V += V_incr;
+          *wTile_V = T1[0][1] - T1[0][3];
+          wTile_V += V_incr;
+          *wTile_V = T1[1][0] - T1[1][2];
+          wTile_V += V_incr;
+          *wTile_V = T1[1][1] + T1[1][2];
+          wTile_V += V_incr;
+          *wTile_V = T1[1][2] - T1[1][1];
+          wTile_V += V_incr;
+          *wTile_V = T1[1][1] - T1[1][3];
+          wTile_V += V_incr;
+          *wTile_V = T1[2][0] - T1[2][2];
+          wTile_V += V_incr;
+          *wTile_V = T1[2][1] + T1[2][2];
+          wTile_V += V_incr;
+          *wTile_V = T1[2][2] - T1[2][1];
+          wTile_V += V_incr;
+          *wTile_V = T1[2][1] - T1[2][3];
+          wTile_V += V_incr;
+          *wTile_V = T1[3][0] - T1[3][2];
+          wTile_V += V_incr;
+          *wTile_V = T1[3][1] + T1[3][2];
+          wTile_V += V_incr;
+          *wTile_V = T1[3][2] - T1[3][1];
+          wTile_V += V_incr;
+          *wTile_V = T1[3][1] - T1[3][3];
+        }
+      }
+    }
+  }
+}
+
+void WinogradConvolution3::Sgemm(const size_t batch_size, const float* weights,
+                                 const size_t input_channels,
+                                 const size_t output_channels) {
+#ifdef USE_MKL
+
+  /*
+   void cblas_sgemm_batch (const CBLAS_LAYOUT Layout, const CBLAS_TRANSPOSE*
```

Comment:
What is this comment?

Reply:
This comment is the commentated signature of the batched sgemm api in MKL.

You won't find this signature on your platform unless you installed MKL. We can get rid of this comment, but  anyone who will have to maintain this method will have to read it at some point.